src/storage/src/hummock/compactor/compaction_executor.rs

```diff
@@ -13,72 +13,56 @@
 // limitations under the License.
 
 use std::future::Future;
-use std::pin::Pin;
 
-use crate::hummock::compactor::CompactOutput;
-use crate::hummock::{HummockError, HummockResult};
+use futures::future::RemoteHandle;
+use futures::FutureExt;
 
-type CompactionRequest = Pin<Box<dyn Future<Output = ()> + Send + 'static>>;
+use crate::hummock::compactor::CompactOutput;
+use crate::hummock::HummockResult;
 
 /// `CompactionExecutor` is a dedicated runtime for compaction's CPU intensive jobs.
 pub struct CompactionExecutor {
-    requests: tokio::sync::mpsc::UnboundedSender<CompactionRequest>,
-    // TODO: graceful shutdown
+    /// Runtime for compaction tasks.
     #[cfg(not(madsim))]
-    _runtime_thread: std::thread::JoinHandle<()>,
+    runtime: &'static tokio::runtime::Runtime,
 }
 
 impl CompactionExecutor {
     #[cfg(not(madsim))]
     pub fn new(worker_threads_num: Option<usize>) -> Self {
-        let (tx, mut rx) = tokio::sync::mpsc::unbounded_channel();
+        let runtime = {
+            let mut builder = tokio::runtime::Builder::new_multi_thread();
+            builder.thread_name("risingwave-compaction");
+            if let Some(worker_threads_num) = worker_threads_num {
+                builder.worker_threads(worker_threads_num);
+            }
+            builder.enable_all().build().unwrap()
+        };
+
         Self {
-            requests: tx,
-            _runtime_thread: std::thread::spawn(move || {
-                let mut builder = tokio::runtime::Builder::new_multi_thread();
-                if let Some(worker_threads_num) = worker_threads_num {
-                    builder.worker_threads(worker_threads_num);
-                }
-                let runtime = builder.enable_all().build().unwrap();
-                runtime.block_on(async {
-                    while let Some(request) = rx.recv().await {
-                        tokio::spawn(request);
-                    }
-                });
-            }),
+            // Leak the runtime to avoid runtime shutting-down in the main async context.
+            // TODO: may manually shutdown the runtime gracefully.
```

Comment:
Is it possible to drop a leaked reference once it's not needed anymore (like graceful shutdown)?

cc. @yezizp2012

Reply:
> Is it possible to drop a leaked reference once it's not needed anymore (like graceful shutdown)?

We can, but not that elegant. I'm considering whether this is necessary: all our tasks should first exit gracefully and release their resources, then the runtime should be clean and it's okay to directly `exit` the process without shutting it down manually.

If the manual shutdown is really necessary, we can refactor this with `Option<Runtime>` and call `shutdown_background` after taking it in `impl Drop`.

cc @Little-Wallace